### ISSUE TITLE
The type 'object{string:'string'}' is more general than the declared return type 'stdClass'

### ISSUE THREAD

**Issue by BenMorel** [vimeo/psalm#8187]
This false positive appeared in `4.24.0`:

```php
function test(): stdClass
{
    return (object) [
        'string' => 'string'
    ];
}
```

https://psalm.dev/r/22e82d7a19

> INFO: [LessSpecificReturnStatement](https://psalm.dev/129) - 5:12 - The type 'object{string:'string'}' is more general than the declared return type 'stdClass' for test

> INFO: [MoreSpecificReturnType](https://psalm.dev/070) - 3:18 - The declared return type 'stdClass' for test is more specific than the inferred return type 'object{string:'string'}'

Even though the cast from `array` to `object` yields an `stdClass`: https://3v4l.org/PkHjR

**Comment by niconoe-** [comment]
This is caused in every direction, like this other snippet: https://psalm.dev/r/5dd2ab19c9

> ERROR: [ArgumentTypeCoercion](https://psalm.dev/193) - 9:6 - Argument 1 of test expects stdClass, parent type object{test:'string'} provided

**Comment by orklah** [comment]
Yeah. The casting could be made more subtle in order to describe the type `stdClass&object{string:'string'}` which is a TNamedObject with extra_types. 

If someone wants to make a PR, it's handled here: https://github.com/vimeo/psalm/blob/e0acf22e402d45d1aebb75f4365ebf701fabce2f/src/Psalm/Internal/Analyzer/Statements/Expression/CastAnalyzer.php#L209

**Comment by edudobay** [comment]
@orklah I tried following the way you suggested and it worked — but I stumbled on another problem due to `stdClass` "testing positive" for existence of any property. I've opened a draft PR for suggestions: https://github.com/vimeo/psalm/pull/8194

**Comment by sapphirecat** [comment]
I have just observed this in 5.0.0-rc1.  I have been lackadaisical about running Psalm, so I'm not sure if it appears earlier in the 5.x betas.

**Comment by niconoe-** [comment]
Is there any news about this bug? If I understand well, @edudobay provided a PR on the 4.x version but due to the release of Psalm 5.x, the PR was closed and nothing have been tried again on the 5.x version.
It looks like this issue was almost closed but now it's back to its original state :disappointed: .

Do you need any help on reproducing the PR @edudobay provided but on 5.x?

**Comment by orklah** [comment]
the PR was unfinished. Potentially, nothing would prevent submitting it again and taking review into account for 5.x

**Comment by edudobay** [comment]
Hi, I haven't been able to work on this since my initial draft. I'll come back to this on March, if no one else starts working on it first.

**Comment by BenMorel** [comment]
@edudobay Looking forward to it!
